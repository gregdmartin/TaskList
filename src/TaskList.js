import React from 'react';
import logo from './logo.svg';
import './TaskList.css';

function TaskList() {
  return (
    <div className="TaskList">
      <header className="TaskList-header">
        <img src={logo} className="TaskList-logo" alt="logo" />
        <p>
          Edit <code>src/App.js</code> and save to reload.
        </p>
        <a
          className="TaskList-link"
          href="https://reactjs.org"
          target="_blank"
          rel="noopener noreferrer"
        >
          Learn React
        </a>
      </header>
    </div>
  );
}

export default TaskList;
