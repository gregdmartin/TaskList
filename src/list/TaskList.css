.TaskList {
  margin-left: 4vmin;
}

.TaskList ul {
  padding-left: inherit;
  width: 20%;
}

.TaskList .task-item {
  list-style: none;
}

.TaskList-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}
