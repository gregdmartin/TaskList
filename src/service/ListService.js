import Ajax from '../common/Ajax';

export const getLists = async () => {
  return Ajax.fetchAjax('api/listing/list?asXhr=true');
};
