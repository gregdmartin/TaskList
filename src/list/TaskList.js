import React from 'react';
import './TaskList.css';
import List from './List';
import { getLists } from '../service/ListService';

/**
 * Controls the entire task list, which includes making nested lists.
 */
class TaskList extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      error: null,
      isLoaded: false,
      lists: []
    };
  }

  /**
   * This method is called by default when it's loaded. This is like an 'onready' event handler.
   * Here, we ask the server for all of the lists we can get. Currently, lists are global.
   * TODO: partition lists by user.
   */
  async componentDidMount() {
    var result, error;
    try {
      result = await getLists();
    } catch(res) {
      error = res.error;
    }

    if(error) {
      this.setState({
        error: error,
        isLoaded: true
      });
    } else {
      if(result && result.lists) {
        let listObjects = result.lists.map((list, i) => {
          return <List key={i} id={list._id} title={list._title} items={list._items} />;
        });
        this.setState({
          isLoaded: true,
          lists: listObjects
        });
      }
    }
  }

  render() {
    let body;
    const {error, isLoaded} = this.state;
    if(!isLoaded) {
      body = <p>Loading...</p>;
    } else if(error) {
      if(error.error) {
        body = <h3>Error: {error.error}</h3>;
      } else {
        body = <h3>An unknown error occurred</h3>
      }
    } else if(this.state.lists.length === 0) {
      body = <h3>No lists! You should create one somehow</h3>
    } else {
      body = this.state.lists;
    }

    return (
      <div className="TaskList">
        <h1>Your Lists</h1>
        {body}
      </div>
    );
  }
}

export default TaskList;
