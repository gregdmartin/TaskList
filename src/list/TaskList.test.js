//https://dev.to/jhotterbeekx/testing-asynchronous-lifecycle-methods-with-jest-13jo


import React from 'react';
import ReactDOM from 'react-dom';
import {shallow, mount} from 'enzyme';

import TaskList from '../list/TaskList';
import { fail } from 'assert';
import { getLists } from '../service/ListService';

jest.mock('../service/ListService');

const fake = {lists: [
  {'_id': 1, '_title': 'test list', '_items': [{'shortDesc': 'something', 'checked': true}]}
]};

describe('TaskList', () => {

  /*it('renders as loading when rendered without mounting', () => {
    const wrapper = shallow(<TaskList />);
    expect(wrapper).toMatchSnapshot();
  });*/

  it('sets error to true if an error occurs', async () => {
    getLists.mockImplementation(() => {
      return Promise.reject({error: new Error('blah')});
    });
    
    const wrapper = mount(<TaskList />);
    await wrapper.instance().componentDidMount();
    
    expect(wrapper.instance().state.isLoaded).toBe(true);
    expect(wrapper.instance().state.lists.length).toEqual(0);
    expect(wrapper.instance().state.error.message).toEqual('blah');
  });

  it('renders lists after they load', async () => {
    getLists.mockImplementation(() => {
      return new Promise((resolve, reject) => resolve(fake));
    });
    
    const wrapper = mount(<TaskList />);
    await wrapper.instance().componentDidMount();
    wrapper.instance().forceUpdate();
    
    expect(wrapper.instance().state.isLoaded).toBe(true);
    expect(wrapper.instance().state.lists.length).toEqual(1)
  });
});
