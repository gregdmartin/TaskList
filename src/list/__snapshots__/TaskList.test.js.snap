// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskList /> render() renders List components for each present list 1`] = `
<div
  className="TaskList"
>
  <h1>
    Your Lists
  </h1>
  <List
    id="1"
    items={Array []}
    key="1"
    title="list title"
  />
</div>
`;

exports[`<TaskList /> render() renders a canned error message when an error object is set but without a messsage 1`] = `
<div
  className="TaskList"
>
  <h1>
    Your Lists
  </h1>
  <h3>
    An unknown error occurred
  </h3>
</div>
`;

exports[`<TaskList /> render() renders a simple message if there are no lists 1`] = `
<div
  className="TaskList"
>
  <h1>
    Your Lists
  </h1>
  <h3>
    No lists! You should create one somehow
  </h3>
</div>
`;

exports[`<TaskList /> render() renders an error message when state.error has a value 1`] = `
<div
  className="TaskList"
>
  <h1>
    Your Lists
  </h1>
  <h3>
    Error: 
    test error
  </h3>
</div>
`;

exports[`<TaskList /> render() renders as loading for the initial state 1`] = `
<div
  className="TaskList"
>
  <h1>
    Your Lists
  </h1>
  <p>
    Loading...
  </p>
</div>
`;
